### DISCUSSION TITLE
Review Club: p2p - Tracing in Bitcoin Core

### DISCUSSION THREAD

**Discussion by yuvicc** [Bitshala/BitcoinCore-PR-Review-Club#112] [PR Reviews]
## Session Details
Date: 19-09-2024
Time: IST 20:00 (UTC 14:30)
Link: [PR #25832](https://github.com/bitcoin/bitcoin/pull/25832)
Difficulty: Medium

## Motivation:
Tracing typically involves capturing and recording information about the system to debug or perform a real-time analysis of the system. It helps identify bugs in the system. When do we add print messages to the code to debug it? We are tracing the program with our own “tracing system”. The debug messages added to the code are static tracepoints, and our “trace system” might send the debug messages to the program output log file (debug.log in Bitcoin). Well, using print statements is a popular way of doing that, but it does not scale as every time we need to trace some part of your code, we need to add some debug statements and build it again. We don’t have to add debug statements every time we need to trace some part of the code because we have built-in tools and frameworks that can do for us, interesting isn’t it? For example, we can collect the data for tracing, which can be done through kprobes, tracepoints, uprobes, dtrace probes/usdt and many more. [Per-peer message capture](https://github.com/bitcoin/bitcoin/pull/19509) was the main motivation towards tracing in Bitcoin core.

## Notes:
The PRs add 5 new Tracepoints which leverage USDT available in Linux. USDT gives lots of flexibility without invasive logging code. It’s more flexible than ad-hoc logging code, allowing one to instrument many different aspects of the system without having to enable per-system logging. It adds three tracepoints:

- established connections with `net:inbound_connection` and `net:outbound_connection`
- closed connections (both closed by the peer or by us) with `net:closed_connnection`
- inbound connections that we choose to evict with `net:evicted_inbound_connection`
- connections that are misbehaving and punished with `net:misbehaving_connection`

## Pre-requisites:
[PR #19509](https://github.com/bitcoin/bitcoin/pull/19509)
[PR #19866](https://github.com/bitcoin/bitcoin/pull/19866)
[PR #22006](https://github.com/bitcoin/bitcoin/pull/22006)
[Discussion on USDT](https://github.com/bitcoin/bitcoin/issues/20981)
Some materials on [USDT](https://leezhenghui.github.io/linux/2019/03/05/exploring-usdt-on-linux.html)
[eBPF](https://www.collabora.com/news-and-blog/blog/2019/04/05/an-ebpf-overview-part-1-introduction/)

## Questions:

- Why does Bitcoin Core include tracing support when we already have the `debug.log` file?
- What do uprobes and kprobes mean to you?
- What does eBPF mean?
- What can some real-world usage of Tracing in Bitcoin Core be?

## Learnings:
P2P
tracing
ebpf
usdt(not stablecoin 😅)

**Comment by paperpsych** [comment]
@VirdiJatin 
![Bitshala RC (190924)](https://github.com/user-attachments/assets/86f172f1-72ef-4d68-9c99-e28c2bee7dd0)

**Comment by yuvicc** [comment]
Looks good to me!! @paperpsych
